### PR TITLE
Always enable listener.ContinueOnListenerFiltersTimeout

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

Always enable continue_on_listener_filter timeout, so that we do not need to distinguish Listener filter timeout nil with 0.

Actually, this field only take effect when timeout set.

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 1]
```diff
@@ -1301,7 +1302,9 @@ func buildListener(opts buildListenerOpts, trafficDirection core.TrafficDirectio
 
 		if opts.proxy.Type != model.Router {
 			res.ListenerFiltersTimeout = opts.push.Mesh.ProtocolDetectionTimeout
-			if res.ListenerFiltersTimeout != nil {
+			// if timeout is not disabled, set ContinueOnListenerFiltersTimeout
+			// if timeout = nil, a default 15s will be applied.
+			if res.ListenerFiltersTimeout == nil || !(res.ListenerFiltersTimeout.GetNanos() == 0 && res.ListenerFiltersTimeout.GetSeconds() == 0) {
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/listener.go:1307]: I know it was there earlier but do we have to check this at all? Why do not we set `ContinueOnListenerFiltersTimeout` by default?

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1307]: Do you mean always make ContinueOnListenerFiltersTimeout = true even no timeout is set? From the doc, there is no harmful effect said

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1307]: Yes. I think that simplifies this and Envoy will use that flag only when listener filter times out.

**Review by ramaraochavali — APPROVED** [review]

**Comment by TommyTu** [comment]
Hey, I am suspecting that this PR changed the behavior of gateway listener implicitly:

Prior to this change, we didn't set ContinueOnListenerFiltersTimeout for gateway listener, and the ListenerFiltersTimeout will be the 15s by default(envoy's default value).

With this change, we set ContinueOnListenerFiltersTimeout always to True for gateway listener, but the ListenerFiltersTimeout will still be the 15s(envoy's default value), due to following logic:

```
		if opts.proxy.Type != model.Router {
			res.ListenerFiltersTimeout = opts.push.Mesh.ProtocolDetectionTimeout
		}
```


Just want to make sure if it is expected change introduced by this PR?

**Comment by hzxuzhonghu** [comment]
I think this code snippet is in `buildGatewayListener`, so actually opts.proxy.Type = model.Router

**Comment by TommyTu** [comment]
> I think this code snippet is in `buildGatewayListener`, so actually opts.proxy.Type = model.Router

Right and this is the problem:
Since we will never assign ListenerFiltersTimeout to gateway listener(we check opts.proxy.Type != model.Router ), the ListenerFiltersTimeout will be envoy's default value([15s](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener.proto)). 

However, in this PR we are assigning `ContinueOnListenerFiltersTimeout` to `true` for gateway listener, which changed its behavior.

**Comment by howardjohn** [comment]
opened https://github.com/istio/istio/pull/47970

**Pull request merged**